Support llvm-mingw toolchain

ied206 opened the pull request:
## Summary

This PR contains patches to make zlib-ng compilable on [llvm-mingw](https://github.com/mstorsjo/llvm-mingw) toolchain.

### Why not build zlib-ng with MinGW-w64 GCC, where support is golden?

MinGW-w64 does not support building arm64 PE files, which llvm-mingw does support.

## PR Details

To make zlib-ng compilable on llvm-mingw, two patches are required.

1. `CMakeList.txt`: Add `-Wno-pedantic-ms-format` only if a toolchain is MinGW GCC.
    - llvm-mingw does not support this argument and causes a warning.
    - This warning will be treated as an error, preventing linking of .dll or .exe.
2. `arch/arm/neon_intrins.h`: Include `arm_neon.h` instead of `arm64_neon.h`.
     - `arm64_neon.h` is MSVC-only header.
     - POSIX GCC, Clang does not have `arm64_neon.h` but `arm_neon.h` on aarch64. MinGW-style clang is also same.
    - Added `MINGW32`, `MINGW64` macro check to make sure not break existing Windows toolchains.

### Reproduce

Download [llvm-mingw.cmake](https://github.com/zlib-ng/zlib-ng/files/12567922/llvm-mingw.cmake.txt), and run cmake command in Powershell (or any other Windows shell).

```pwsh
# Generate Makefiles
> cmake .. -G "MinGW Makefiles" `
"-DCMAKE_MAKE_PROGRAM=<LLVM_MINGW_ROOT_PATH>\bin\mingw32-make.exe" `
"-DCMAKE_TOOLCHAIN_FILE=<DOWNLOAD_PATH>\llvm-mingw.cmake" `
"-DCMAKE_SYSTEM_PROCESSOR=aarch64" `
"-DLLVM_MINGW=<LLVM_MINGW_ROOT_PATH>" 
# Build zlib-ng
> cmake --build . --config Release
```

`CMAKE_SYSTEM_PROCESSOR` can be one of i686, x86_64 and aarch64.

### Considered but not included

If I use `MinGW Makefiles` generator to build zlib-ng, `libzlib-ng2.dll` will be produced. This is different from the `zlib-ng2.dll` MSVC build and the traditional `zlib1.dll` naming convention. Is this appending `lib` prefix an intended behavior?

I found out that adding this line into the CMakeList.txt prevents this. But I did not include it here because I was not sure about the intent.

```
if(MINGW)
...
   set(CMAKE_SHARED_LIBRARY_PREFIX "")
...
endif()
```

ied206 committed:
Support llvm-mingw toolchain

zlib-ng requires some patches to make it compilable on LLVM-mingw.
1. Add -Wno-pedantic-ms-format only if a toolchain is MinGW GCC.
- llvm-mingw does not support it, causing build to break.
2. Include arm_neon.h instead of arm64_neon.h (aarch64 only).
- arm64_neon.h is MSVC only.
- GCC, Clang does not have arm64_neon.h but arm_neon.h on aarch64.
- Also applied to configure and detect-instrinsics.cmake

nmoinvaz commented on arch/arm/neon_intrins.h:
```diff
@@ -1,8 +1,14 @@
 
-#if defined(_M_ARM64) || defined(_M_ARM64EC)
-#  include <arm64_neon.h>
+#if (defined(_M_ARM64) || defined(_M_ARM64EC))
```
These includes are also part of tests in CMake and configure build scripts.

I'm thinking it would be better to use a `check_include_file`, but @mtl1979 might have ideas on that.

nmoinvaz replied:
Actually it is only used in source in one place. So maybe it is not needed. But those other instances should be updated. But honestly it kind of seems to me like a toolchain bug.

mtl1979 replied:
The problem with check_include_file() is that it might find a header for different target when toolchain supports multiple architectures as is with Windows API.

mtl1979 commented:
If something is specific to single configuration, it should be added to toolchain file, not CMakeLists.txt... This makes sure it is used from very start of the root CMakeLists.txt. We can include customized toolchain files in this repository.

ied206 commented:
> If something is specific to single configuration, it should be added to toolchain file, not CMakeLists.txt... This makes sure it is used from very start of the root CMakeLists.txt. We can include customized toolchain files in this repository.

I will add llvm-mingw CMake custom toolchain files.

ccawley2011 commented on arch/arm/neon_intrins.h:
```diff
@@ -1,8 +1,14 @@
+#  else
+/* In MSVC, including arm64_neon.h is necessary */
+#     include <arm64_neon.h>
+#  endif
```
Since `arm64_neon.h` is specific to MSVC, it might be better to check for that specifically.

```suggestion
#if defined(_MSC_VER) && (defined(_M_ARM64) || defined(_M_ARM64EC))
#  include <arm64_neon.h>
```

ied206 replied:
> Since `arm64_neon.h` is specific to MSVC, it might be better to check for that specifically.

Thanks for advising. I have applied your suggestion in the latest commit.

nmoinvaz commented:
@Dead2 did you see @ccawley2011's comment?

Dead2 commented:
> @Dead2 did you see @ccawley2011's comment?
Thanks, I don't think I did. Awaiting the next revision. 👍

nmoinvaz commented on CMakeLists.txt:
```diff
@@ -236,7 +236,11 @@ elseif(CMAKE_C_COMPILER_ID MATCHES "GNU" OR CMAKE_C_COMPILER_ID MATCHES "Clang")
     endif()
     if(MINGW)
-        list(APPEND WARNFLAGS_DISABLE -Wno-pedantic-ms-format)
+        # Add `-Wno-pedantic-ms-format` only if the toolchain is GCC MinGW-w64
```
What are the warnings you were seeing? Can you list the compiler output? AFAIK, from a search of our GitHub code, we don't use `%I64d` or `%I32d`

ied206 replied:
> What are the warnings you were seeing? Can you list the compiler output? AFAIK, from a search of our GitHub code, we don't use `%I64d` or `%I32d`

clang simply does not recognize the `-Wno-pedantic-ms-format` argument.
So the toolchain always prints the `unknown warning option` warning in builds.
This does not cause any harm in compiling .o files, but it is treated as an error in linking testing binaries.

This is a build log from compiling unmodified zlib-ng-2.1.3 with llvm-mingw.
```
[ 76%] Building C object test/CMakeFiles/maketrees.dir/__/zutil.c.obj
warning: unknown warning option '-Wno-pedantic-ms-format'; did you mean '-Wno-pedantic-macros'? [-Wunknown-warning-option]
1 warning generated.
[ 76%] Linking C executable ..\maketrees.exe
[ 76%] Built target maketrees
[ 76%] Building C object test/CMakeFiles/makecrct.dir/__/tools/makecrct.c.obj
warning: unknown warning option '-Wno-pedantic-ms-format'; did you mean '-Wno-pedantic-macros'? [-Wunknown-warning-option]
1 warning generated.
[ 77%] Linking C executable ..\makecrct.exe
[ 77%] Built target makecrct
[ 78%] Building CXX object _deps/googletest-build/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.obj
error: unknown warning option '-Wno-pedantic-ms-format'; did you mean '-Wno-pedantic-macros'? [-Werror,-Wunknown-warning-option]
mingw32-make[2]: *** [_deps\googletest-build\googletest\CMakeFiles\gtest.dir\build.make:76: _deps/googletest-build/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.obj] Error 1
mingw32-make[1]: *** [CMakeFiles\Makefile2:509: _deps/googletest-build/googletest/CMakeFiles/gtest.dir/all] Error 2
mingw32-make: *** [Makefile:145: all] Error 2
```

nmoinvaz replied:
Here is the original PR for the `-Wno-pedantic-ms-format` change https://github.com/zlib-ng/zlib-ng/pull/1362.

nmoinvaz replied:
Another option would be to check if the compiler supported the flag.
```cmake
check_c_compiler_flag(-Wno-pedantic-ms-format HAVE_NO_PEDANTIC_MS_FORMAT)
if(HAVE_NO_PEDANTIC_MS_FORMAT)
    list(APPEND WARNFLAGS_DISABLE -Wno-pedantic-ms-format)
endif()
```
It might be overkill, idk..

ied206 replied:
I didn't know CMake had such capabilities. I will try checking it.

Dead2 replied:
Was this change also going in or not?

ied206 replied:
Sorry, I forgot to test this last time! Fortunately, it works like a charm on both llvm-mingw and MinGW-w64 GCC.

https://github.com/zlib-ng/zlib-ng/pull/1570/commits/3d14566fa9bc042f88c879d5c2813b738990d632 introduced `check_c_compiler_flag` based `-Wno-pedantic-ms-format` detection.

ied206 commented:
## Status

- Changed `arm64_neon.h` code to check existance of `_MSC_VER` instead of `MINGW32`/`MINGW64`.
   - This change is also applied to `detect-instrinsics.cmake` and `configure`. 
   - I am unsure how `configure` is generated yet, so I did it by hand. If I did wrong, please let me know.
- llvm-mingw toolchain files are being worked.
   - The original toolchain files I used are licensed differently from this project, so I am rewriting it.
   - The current version of the rewrite is failing to read the toolchain root path from `CMakeCache.txt`, unlike its original. Using the environment variable to store the path works, but it is not intuitive enough.
   - Does the zlib-ng project have strict rules about how CMake toolchain files can retrieve the toolchain root path?

ied206 commented:
This is a half-baked cmake toolchain file for llvm-mingw. 
It is not complete (and thus not committed) yet for the reasons I wrote above, but I am uploading the preliminary version for feedback.

[toolchain-llvm-mingw-i686.cmake](https://github.com/zlib-ng/zlib-ng/files/12644686/toolchain-llvm-mingw-i686.cmake.txt)

- It was rewritten to follow zlib-ng mingw-w64 toolchain file conventions and notations.
- It follows the zlib-ng project-wide license.

Dead2 reviewed (changes requested):
I would like the commits squashed before merging, as the two commits provide no value from being separated. I could do this when merging, but preferably not if a toolchain file is added later.

Dead2 commented:
Configure is not generated, it is made manually, so that change is fine.

I am not sure about the toolchain questions, that is an area that I am not very familiar with myself unfortunately.
It is great that you are doing a rewrite to ensure that licensing is correct 👍 

I also approved the CI, so it will tell us whether anything broke from these changes so far (but I do not expect any).

nmoinvaz commented on CMakeLists.txt:
```diff
@@ -236,7 +236,11 @@ elseif(CMAKE_C_COMPILER_ID MATCHES "GNU" OR CMAKE_C_COMPILER_ID MATCHES "Clang")
-        list(APPEND WARNFLAGS_DISABLE -Wno-pedantic-ms-format)
+        # Add `-Wno-pedantic-ms-format` only if the toolchain is GCC MinGW-w64
+        # llvm-mingw does not support '-Wno-pedantic-ms-format'
+        if (CMAKE_C_COMPILER_ID MATCHES "GNU")
```
Extra space after `if`.

ied206 replied:
I applied your review to [e0dd69ad3d52d4b0f2a2952cbe44df5bd4466c4d](https://github.com/zlib-ng/zlib-ng/pull/1570/commits/9708ad83a50d352813c43cf37e452ce9cbd65f43) (Related commit squashed to one commit).

nmoinvaz commented:
> This is a half-baked cmake toolchain file for llvm-mingw.

Was your plan to add the toolchain file into our `cmake` directory? Then incorporate it into the GitHub CI?

ied206 commented:
> Was your plan to add the toolchain file into our `cmake` directory? Then incorporate it into the GitHub CI?

My intent was to provide users with a simple way to compile zlib-ng with llvm-mingw. I didn't know it could be incorporated into the GitHub CI. 

Many CIs do not provide llvm-mingw by default, though. Incorporating it would require considering bandwidth usage, so it is not my primary consideration now.

ied206 committed:
Add llvm-mingw CMake toolchain files

ied206 commented:
## Status Report - Done

I finished the requested changes.

- Space after `if` is removed in `CMakeList.txt` (reviewed by @nmoinvaz).
- All changes related to conditional `arm64_neon.h` include are squashed into one commit (reviewed by @Dead2).
- CMake toolchain files, `toolchain-llvm-mingw-*.cmake`, are added.
    - Supports `i686`, `x86_64`, `armv7`, `aarch64` architectures.
    - Tested on Windows and Linux ([llvm-mingw](https://github.com/mstorsjo/llvm-mingw/releases) project provides Windows, Linux and macOS binaries), so it can run on any CI environments.
    - The toolchain location can be passed by one of these:
        1. Add llvm-mingw toolchain root to `$PATH`.
        2. Set `-DLLVM_MINGW_ROOT=<TOOLCHAIN_ROOT>`.
        3. Set `LLVM_MINGW_ROOT` environment variable.

### Examples of using llvm-mingw CMake toolchain file:

- On Windows Powershell (using LLVM_MINGW_ROOT):
```powershell
> mkdir build-aarch64
> cd build-aarch64
> cmake .. -G "MinGW Makefiles" `
"-DCMAKE_MAKE_PROGRAM=C:\Tools\llvm-mingw\bin\mingw32-make.exe" `
"-DCMAKE_TOOLCHAIN_FILE=..\cmake\toolchain-llvm-mingw-aarch64.cmake" `
"-DLLVM_MINGW_ROOT=C:\Tools\llvm-mingw"
> cmake --build . --config Release --parallel 16
```
- On Linux bash (using PATH):
```bash
$ export PATH=$HOME/Tools/llvm-mingw-20230919-ucrt-ubuntu-20.04-x86_64/bin:$PATH
$ mkdir build-aarch64
$ cd build-aarch64
$ cmake .. -G "Unix Makefiles" \
"-DCMAKE_TOOLCHAIN_FILE=../cmake/toolchain-llvm-mingw-aarch64.cmake"
$ cmake --build . --config Release --parallel 16
```

### Appendix: `lib` prefix on DLL files issue

> If I use MinGW Makefiles generator to build zlib-ng, `libzlib-ng2.dll` will be produced. This is different from the `zlib-ng2.dll` MSVC build and the traditional `zlib1.dll` naming convention. Is this appending lib prefix an intended behavior?

> If something is specific to single configuration, it should be added to toolchain file, not CMakeLists.txt... This makes sure it is used from very start of the root CMakeLists.txt. We can include customized toolchain files in this repository.

I have tried putting `set(CMAKE_SHARED_LIBRARY_PREFIX "")` to llvm-mingw toolchain files, but it does not work. Only putting it directly to `CMakeLists.txt` works. My guess is that the value is being overridden by CMakeLists.txt itself. 

Discussion about the `lib` prefix on Windows is not vital on this PR, it can be discussed in a separate issue/PR.

ied206 committed:
Detect -Wno-pedantic-ms-format support in mingw

Use check_c_compiler_flag() feature of CMake to figure out whether this
mingw toolchain has support for -Wno-pedantic-ms-format or not.
This flag will be added only if the toolchain supports it.
Currently, MinGW-w64 GCC has it, and llvm-mingw does not have it.

ied206 commented:
`CMakeLists.txt` now adds `-Wno-pedantic-ms-format` flag on mingw only if the toolchain supports this flag.
It is done by using `check_c_compiler_flag()`, which was proposed by @nmoinvaz.

This new version of `CMakeLists.txt` is tested on both llvm-mingw and MinGW-w64 GCC.

nmoinvaz reviewed (approved):
LGTM. We can add CI tests for cmake toolchain files if Ubuntu ever picks up llvm-mingw package, but I'm not worried about it.

nmoinvaz commented:
@ied206 can you copy those instructions for building with `llvm-mingw` to this Wiki page? https://github.com/zlib-ng/zlib-ng/wiki/Build-tips

ied206 commented:
> @ied206 can you copy those instructions for building with `llvm-mingw` to this Wiki page? https://github.com/zlib-ng/zlib-ng/wiki/Build-tips

I have added build instructions about overall MinGW toolchains including `llvm-mingw`.
I also added a basic explanation of CMake toolchain files for `llvm-mingw` and `MinGW-w64`.

Dead2 reviewed (approved):
LGTM
